runtime/caml/misc.h

```diff
@@ -185,12 +185,25 @@ extern "C" {
 #endif
 
 /* GC timing hooks. These can be assigned by the user. These hooks
-   must not allocate, change any heap value, nor call OCaml code.
-*/
+   must not allocate, change any heap value, nor call OCaml code. They
+   can obtain the domain id with Caml_state->id. These functions must
+   be reentrant. */
 typedef void (*caml_timing_hook) (void);
-extern caml_timing_hook caml_major_slice_begin_hook, caml_major_slice_end_hook;
-extern caml_timing_hook caml_minor_gc_begin_hook, caml_minor_gc_end_hook;
-extern caml_timing_hook caml_finalise_begin_hook, caml_finalise_end_hook;
+extern _Atomic caml_timing_hook caml_major_slice_begin_hook;
+extern _Atomic caml_timing_hook caml_major_slice_end_hook;
+extern _Atomic caml_timing_hook caml_minor_gc_begin_hook;
+extern _Atomic caml_timing_hook caml_minor_gc_end_hook;
+extern _Atomic caml_timing_hook caml_finalise_begin_hook;
+extern _Atomic caml_timing_hook caml_finalise_end_hook;
+
+#ifdef CAML_INTERNALS
+
+#define Call_timing_hook(a) do {                                         \
+    caml_timing_hook h = atomic_load_explicit(&a, memory_order_relaxed); \
+    if (h != NULL) (*h)();                                               \
+  } while (0)
```

Comment:
Is there a strong reason for making this a macro? I would just use a static inline, taking the hook address as parameter.

Reply:
Right, it's better to avoid macros when we can.